File: src/main/java/org/pl/interpreter/visitor/LoopVisitor.java
package org.pl.interpreter.visitor;

import org.pl.interpreter.IInterpreter;
import org.pl.interpreter.exception.VisitorException;
import org.pl.parser.ast.INode;
import org.pl.parser.ast.LoopNode;

import java.util.ArrayList;
import java.util.Collection;


/**
 * Interprets the loop logic.
 */
public class LoopVisitor implements IVisitor {

    @Override
    public boolean matches(INode node) {
        return LoopNode.class == node.getClass();
    }

    @Override
    public Object visit(IInterpreter interpreter, INode node) throws VisitorException {
        var loopNode = (LoopNode) node;
        var result = new ArrayList<Object>(8);
        while (loopNode.getCondition() == null || Boolean.TRUE.equals(interpreter.evalNode(loopNode.getCondition()))) {
            var bodyResult = interpreter.evalNode(loopNode.getBody());

            // I want to prevent very deep lists from occurring. This check prevents that
            // something like [true] becomes [[[true]]]
            if (bodyResult instanceof Collection) {
                result.addAll((Collection<?>) bodyResult);
            } else {
                result.add(bodyResult);
            }
        }
        return result;
    }
}

File: src/main/kotlin/org/leaflang/interpreter/visitor/LoopVisitor.kt
package org.leaflang.interpreter.visitor

import org.leaflang.interpreter.IInterpreter
import org.leaflang.interpreter.result.*
import org.leaflang.interpreter.withStaticScope
import org.leaflang.parser.ast.INode
import org.leaflang.parser.ast.LoopNode

/**
 * Interprets the loop logic.
 */
class LoopVisitor : IVisitor {

    override fun visit(interpreter: IInterpreter, node: INode): IRuntimeResult {
        val loopNode = node as LoopNode
        var result: IRuntimeResult = emptyResult()

        interpreter.withStaticScope("loop") {

            // Do the initialization of the loop
            if (loopNode.init != null) interpreter.interpret(loopNode.init)

            // Run the actual loop
            while (loopNode.condition == null ||
                    interpreter.interpret(loopNode.condition).data?.value == true) {
                val bodyResult = interpreter.interpret(loopNode.body)

                // Break the loop if necessary
                if (bodyResult is BreakRuntimeResult) break

                // Do one step of the loops step expression
                interpreter.interpret(loopNode.step)

                // Other options to break out of a loop
                if (bodyResult is ContinueRuntimeResult) continue
                if (bodyResult is ReturnRuntimeResult) {
                    result = bodyResult
                    break
                }
            }
        }
        return result
    }
}